main.go

```diff
@@ -41,41 +31,39 @@ func main() {
 		return
 	}
 
-	ifaceAndPort := fmt.Sprintf("%v:0", *iface)
-	listener, err := net.Listen("tcp4", ifaceAndPort)
-	if err != nil {
-		log.Fatal(err)
-	}
-
-	pr, pw, _ := os.Pipe()
-	gcChan := make(chan *gctrace, 1)
-	scvgChan := make(chan *scvgtrace, 1)
-
-	parser := Parser{
-		reader:   pr,
-		gcChan:   gcChan,
-		scvgChan: scvgChan,
-	}
-
-	gcvisGraph = NewGraph(strings.Join(flag.Args(), " "), GCVIS_TMPL)
+	subcommand := NewSubCommand(flag.Args())
+	parser := NewParser(subcommand.PipeRead)
+	gcvisGraph := NewGraph(strings.Join(flag.Args(), " "), GCVIS_TMPL)
+	server := NewHttpServer(*iface, *port, &gcvisGraph)
 
-	go startSubprocess(pw, flag.Args())
+	go subcommand.Run()
 	go parser.Run()
+	go server.Start()
 
-	http.HandleFunc("/", indexHandler)
-
-	go http.Serve(listener, nil)
-
-	url := fmt.Sprintf("http://%s/", listener.Addr())
+	url := server.Url()
 	log.Printf("opening browser window, if this fails, navigate to %s", url)
 	browser.OpenURL(url)
 
 	for {
 		select {
-		case gcTrace := <-gcChan:
+		case gcTrace := <-parser.GcChan:
 			gcvisGraph.AddGCTraceGraphPoint(gcTrace)
-		case scvgTrace := <-scvgChan:
+		case scvgTrace := <-parser.ScvgChan:
 			gcvisGraph.AddScavengerGraphPoint(scvgTrace)
+		case output := <-parser.NoMatchChan:
+			fmt.Fprintln(os.Stderr, output)
+		case <-parser.done:
+			if parser.Err != nil {
+				fmt.Fprintf(os.Stderr, parser.Err.Error())
+				os.Exit(1)
+			}
+
+			if subcommand.Err != nil {
```

Comment:
This is a race as there is no synchronisation between the goroutine setting .Err and this one reading it.

A simple solution might be to have subcommnd.Run just print the error, and drop the os.Exit(1) case here.